File: Inboxes/PHP/Web/Plain/src/Storage/Adapters/MongoDB.php
<?php
require_once __DIR__ . '/../DatabaseAdapter.php';

class MongoDB implements DatabaseAdapter {
    private ?MongoDB\Database $db = null;

    public function connect(): void {
        $host = getenv('DB_HOST') ?: 'localhost';
        $port = getenv('DB_PORT') ?: '27017';
        $name = getenv('DB_NAME') ?: 'inboxes';
        $mongoClient = new MongoDB\Client("mongodb://$host:$port");
        $this->db = $mongoClient->selectDatabase($name);
    }

    private function getCollection(string $table): MongoDB\Collection {
        return $this->db->$table;
    }

    private function docToArray($doc): array {
        $arr = $doc->getArrayCopy();
        $arr['id'] = (string)$doc['_id'];
        unset($arr['_id']);
        return $arr;
    }

    public function getAll(string $table): array {
        $docs = $this->getCollection($table)->find([], ['sort' => ['_id' => 1]])->toArray();
        return array_map([$this, 'docToArray'], $docs);
    }

    public function getById(string $table, int|string $id): ?array {
        $doc = $this->getCollection($table)->findOne(['_id' => new MongoDB\BSON\ObjectId((string)$id)]);
        if (!$doc) return null;
        return $this->docToArray($doc);
    }

    public function create(string $table, array $data): int|string {
        $result = $this->getCollection($table)->insertOne($data);
        return (string)$result->getInsertedId();
    }

    public function update(string $table, int|string $id, array $data): bool {
        $result = $this->getCollection($table)->updateOne(
            ['_id' => new MongoDB\BSON\ObjectId((string)$id)],
            ['$set' => $data]
        );
        return $result->getModifiedCount() > 0 || $result->getMatchedCount() > 0;
    }

    public function delete(string $table, int|string $id): bool {
        $result = $this->getCollection($table)->deleteOne(['_id' => new MongoDB\BSON\ObjectId((string)$id)]);
        return $result->getDeletedCount() > 0;
    }

    public function search(string $table, string $field, string $query): array {
        $docs = $this->getCollection($table)->find(
            [$field => new MongoDB\BSON\Regex($query, 'i')],
            ['sort' => ['_id' => 1]]
        )->toArray();
        return array_map([$this, 'docToArray'], $docs);
    }
}

File: Inboxes/Python/Web/Plain/src/storage/adapters/mongodb.py
import os
import re
from bson.objectid import ObjectId
from pymongo import MongoClient
from storage.database_adapter import DatabaseAdapter

class MongoDB(DatabaseAdapter):
    def __init__(self):
        self.db = None
        self.connect()

    def connect(self):
        if self.db:
            return
        client = MongoClient(
            host=os.getenv("DB_HOST", "localhost"),
            port=int(os.getenv("DB_PORT", "27017")),
        )
        self.db = client[os.getenv("DB_NAME", "inboxes")]

    def _serialize(self, doc):
        if doc and "_id" in doc:
            doc["id"] = str(doc.pop("_id"))
        return doc

    def init_table(self, table, columns=None):
        self.connect()
        if table not in self.db.list_collection_names():
            self.db.create_collection(table)

    def get_all(self, table):
        self.connect()
        return [self._serialize(doc) for doc in self.db[table].find({}).sort("_id", 1)]

    def get_by_id(self, table, id):
        self.connect()
        doc = self.db[table].find_one({"_id": ObjectId(id)})
        return self._serialize(doc)

    def create(self, table, data):
        self.connect()
        result = self.db[table].insert_one(data)
        doc = self.db[table].find_one({"_id": result.inserted_id})
        return self._serialize(doc)

    def update(self, table, id, data):
        self.connect()
        result = self.db[table].update_one(
            {"_id": ObjectId(id)}, {"$set": data}
        )
        return result.modified_count > 0

    def delete(self, table, id):
        self.connect()
        result = self.db[table].delete_one({"_id": ObjectId(id)})
        return result.deleted_count > 0

    def search(self, table, field, query):
        self.connect()
        pattern = re.compile(re.escape(query), re.IGNORECASE)
        return [
            self._serialize(doc)
            for doc in self.db[table].find({field: {"$regex": pattern}}).sort("_id", 1)
        ]

    def close(self):
        if self.db:
            self.db.client.close()
